### PR TITLE
New structure

### PR DESCRIPTION
**Pull request by samuelWilliams99**
This PR is now functionally complete!
I've redesigned this lib to treat notifications as objects, with properties, callbacks, etc. (Like vgui objects)
I've also implementing some helper functions for quick use of simple notifications, e.g. ` CFCNotifications.sendSimple` and `CFCNotifications.sendImportantSimple`

I'll add some example code to this later, if you want to add suggestions to the lib, now is the time to do it!

### CODE REVIEW

**Comment by samuelWilliams99** [comment]
This is currently untested, the re-design has never been run. Don't be surprised if you try run it and it doesn't work.

**Comment by samuelWilliams99** [comment]
Jenkin's is now only complaining about unused variables in my incomplete code.

**Comment by samuelWilliams99** [comment]
I have no remorse.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,173 @@
...
+
+## Overview
+This library acts as a framework for notifications of any functionality to created, managed and removed.  
+Notifications are treated as objects with methods and fields, these objects have a type which defines how they look, behave, etc. and they can be created and displayed from server or client.  
+One can also easily define their own notification types, similar to how you would create a vgui element.
+Each notification object has an ID, simply a string.  
+This allows the notification to be blocked (temporarily or permanently) by specific clients (if permitted) and for networking of notification events.
+
+## Client
+Clients can press F3 to release their mouse and interact with notifications.
```
Comment (sarahsturgeon) [inline README.md]: Can they also use the context menu or open up chat?

File: lua/cfc_notifications/client/cl_dnotification.lua [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+function PANEL:Think()
+    local ct = SysTime()
+    local pt = self._prevThinkTime or 0
+    local dt = ct - pt
+    self._prevThinkTime = ct
+    if dt > 0.5 then
+        -- Too much time passed between paint, probably wasn't visible
+        dt = 0
+    end
+    self._thinkDeltaTime = dt
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotification.lua]: Can we get some better variable names here?

File: lua/cfc_notifications/client/cl_dnotification.lua [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+function PANEL:Paint( w, h )
+    solidColorPaint( self, w, h )
+    if not self._showTimer then return end
+
+    local frac = self._curTime / self._maxTime
+    local barHeight = 4
+    surface.SetDrawColor( Color( 180, 180, 180, 150 ) )
+    surface.DrawRect( 0, h-barHeight, w, barHeight )
+    surface.SetDrawColor( Color( 240, 40, 40 ) )
+    surface.DrawRect( 0, h-barHeight, w * frac, barHeight )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotification.lua]: Spaces around operators (`-`) and may want to do the calculations outside of the draw calls

File: lua/cfc_notifications/client/cl_dnotification.lua [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    local remaining = math.Clamp( self._maxTime - self._curTime, 0, 100000 )
+    local remainingStr = secondsAsTime( math.ceil( remaining ) )
+    draw.DrawText( remainingStr, "CFC_Notifications_Mono", w - 44, h - 16, Color( 180, 180, 180 ) )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotification.lua]: Same comment about maybe doing the calculations outside of the Draw call.

Also, might be good to assign the color to a variable outside of here (CFC color lib?)

File: lua/cfc_notifications/client/cl_dnotification.lua [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    local data = { ... }
+    function btn:DoClick()
+        if not this:GetButtonsDisabled() and this[cbName] then
+            this[cbName]( this, unpack( data ) )
+        end
+    end
+    function btn:Think()
+        local x, y = self:LocalCursorPos()
+        local w, h = self:GetSize()
+        local isHovered = x >= 0 and x <= w and y >= 0 and y <= h -- Not using panel:IsHovered as that often gets it wrong due to using vgui.GetHoveredPanel()
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotification.lua]: Could we get the comment above this line and explain what the actual issue is, and why this method solves it?

File: lua/autorun/server/sv_cfc_n_init.lua [hunk 6]
```diff
@@ -0,0 +1,16 @@
+function addFiles( dir )
```
Comment (samuelWilliams99) [inline lua/autorun/server/sv_cfc_n_init.lua]:
```suggestion
local function addFiles( dir )
```

File: lua/cfc_notifications/client/cl_dnotification.lua [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        offset = offset + w
+        btn:SetPos( barWidth - offset, ( 20 - h ) / 2 )
+
+        if k ~= #btnTexts then
+            local splitBar = vgui.Create( "DLabel", bar )
+            splitBar:SetText( "|" )
+            splitBar:SetTextColor( Color( 200, 200, 200 ) )
+            splitBar:SizeToChildren()
+            splitBar:SetPos( barWidth - ( offset + 4 ), 0 )
+            offset = offset + 5
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotification.lua]: Let's extract these magic numbers out somewhere with proper names. May want to do the calcs outside that explains what it does.

```lua
local widthWithOffset = blah
```

File: lua/cfc_notifications/client/cl_dnotificationbutton.lua [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    self.underlineWeight = 1
+    self.clickAnimationLength = 1
+    self.clickAnimState = 0
+    self:SetTextColor( Color( 255, 255, 255 ) )
+    self:SetBackgroundColor( Color( 100, 100, 100 ) )
+end
+
+function PANEL:Think()
+    local disabled = self:GetDisabled()
+    local tc = self.textCol
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotificationbutton.lua]: tc -> textColor

File: lua/cfc_notifications/client/cl_dnotificationbutton.lua [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        end
+    end
+
+    local s = math.sin( self.animState * ( math.pi / 2 ) )
+
+    surface.SetDrawColor( self:GetBackgroundColor() )
+    surface.DrawRect( 0, h - uWeight - 4, w, uWeight )
+
+    surface.SetDrawColor( self.textCol )
+    surface.DrawRect( w * ( 1 - s ) * 0.5, h - uWeight - 4, w * s, uWeight )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_dnotificationbutton.lua]: Very, very hard to follow this part. I would struggle to modify this in the future with all of these inline calculations

File: lua/cfc_notifications/client/cl_net.lua [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    notif:Send()
+end )
+
+CFCNotifications._serverNotificationIDs = {}
+
+net.Receive( "CFC_NotificationExists", function( len )
+    local id = net.ReadString()
+    local exists = net.ReadBool()
+    local hasValue = table.HasValue( CFCNotifications._serverNotificationIDs, id )
+    if exists and not hasValue then
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_net.lua:20]:
```suggestion

    if exists and not hasValue then
```

File: lua/cfc_notifications/client/cl_save.lua [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+function CFCNotifications.Base:Ignore( permanent )
+    local id = self:GetID()
+
+    if permanent then
+        CFCNotifications._permIgnores[id] = true
+        CFCNotifications.saveIgnores()
+    else
+        CFCNotifications._tempIgnores[id] = true
+    end
+    CFCNotifications._reloadIgnoredPanels()
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_save.lua:23]:
```suggestion

    CFCNotifications._reloadIgnoredPanels()
```

File: lua/cfc_notifications/client/cl_save.lua [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+            table.insert( out, k )
+        end
+    end
+    for k, v in pairs( CFCNotifications._serverNotificationIDs ) do
+        if not CFCNotifications._tempIgnores[v] and not CFCNotifications._permIgnores[v] and not table.HasValue( out, v ) then
+            table.insert( out, v )
+        end
+    end
+    return out
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_save.lua]: Can we get some whitespace in here?

```suggestion
function CFCNotifications.getUnignored()
    local out = {}
    
    for k, v in pairs( CFCNotifications.Notifications ) do
        if not CFCNotifications._tempIgnores[k] and not CFCNotifications._permIgnores[k] and v:GetCloseable() and v:GetIgnoreable() then
            table.insert( out, k )
        end
    end
    
    for k, v in pairs( CFCNotifications._serverNotificationIDs ) do
        if not CFCNotifications._tempIgnores[v] and not CFCNotifications._permIgnores[v] and not table.HasValue( out, v ) then
            table.insert( out, v )
        end
    end
    
    return out
end
```

File: lua/cfc_notifications/client/cl_save.lua [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    end
+    panels[2]:Clear()
+    for k, v in ipairs( CFCNotifications.getPermIgnored() ) do
+        panels[2]:AddItem( v )
+    end
+    panels[3]:Clear()
+    for k, v in ipairs( CFCNotifications.getUnignored() ) do
+        panels[3]:AddItem( v )
+    end
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_save.lua]:
```suggestion
function CFCNotifications._reloadIgnoredPanels()
    if #panels == 0 then return end
    
    panels[1]:Clear()
    
    for k, v in ipairs( CFCNotifications.getTempIgnored() ) do
        panels[1]:AddItem( v )
    end
    
    panels[2]:Clear()
    
    for k, v in ipairs( CFCNotifications.getPermIgnored() ) do
        panels[2]:AddItem( v )
    end
    
    panels[3]:Clear()
    
    for k, v in ipairs( CFCNotifications.getUnignored() ) do
        panels[3]:AddItem( v )
    end
end
```

File: lua/cfc_notifications/client/cl_save.lua [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    list:SetMultiSelect( false )
+
+    function list:OnRowSelected( idx, row )
+        timer.Simple( 0, function()
+            if not row:IsValid() then return end
+            self:ClearSelection()
+            local id = row:GetColumnText( 1 )
+            self:RemoveLine( idx )
+            onLeft( id )
+            CFCNotifications._reloadIgnoredPanels()
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_save.lua]:
```suggestion
            if not row:IsValid() then return end
            
            self:ClearSelection()
            
            local id = row:GetColumnText( 1 )
            
            self:RemoveLine( idx )
            onLeft( id )
            
            CFCNotifications._reloadIgnoredPanels()
```

File: lua/cfc_notifications/client/cl_save.lua [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        end )
+    end
+
+    function list:OnRowRightClick( idx, row )
+        self:ClearSelection()
+        local id = row:GetColumnText( 1 )
+        self:RemoveLine( idx )
+        onRight( id )
+        CFCNotifications._reloadIgnoredPanels()
+    end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_save.lua]: Whitespace, brother, we have plenty of space

File: lua/cfc_notifications/client/cl_settings.lua [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+    for k, v in pairs( CFCNotifications._settingsTemplate ) do
+        if v.name == name then
+            local val = typeGetters[v.type]( cVar )
+            if not v.noCache then
+                CFCNotifications._settingCache[name] = val
+            end
+            return val
+        end
+    end
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]: More whitespace please

File: lua/cfc_notifications/client/cl_settings.lua [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+    for k, setting in pairs( CFCNotifications._settingsTemplate ) do
+        if setting.type ~= "action" then
+            local val = "cfc_notifications_" .. setting.name
+            local cv = GetConVar( val )
+            typeSetters[setting.type]( cv, setting.default )
+        end
+    end
+    print( "Reverted CFCNotifications to default, reloading..." )
+    CFCNotifications.reload()
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]:
```suggestion
function CFCNotifications.resetSettings()
    for k, setting in pairs( CFCNotifications._settingsTemplate ) do
        if setting.type ~= "action" then
            local val = "cfc_notifications_" .. setting.name
            local cv = GetConVar( val )
            
            typeSetters[setting.type]( cv, setting.default )
        end
    end
    
    print( "Reverted CFCNotifications to default, reloading..." )
    
    CFCNotifications.reload()
end
```

File: lua/cfc_notifications/client/cl_settings.lua [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        elseif table.HasValue( fVals, val ) then
+            return true, "0"
+        else
+            return false, "Invalid boolean value"
+        end
+    end,
+    string = function( data, val )
+        return true, val
+    end,
+}
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]: Not a fan of inlined functions with such depth, could we extract these somewhere else and then plug them in here? Some validator file or something?

File: lua/cfc_notifications/client/cl_settings.lua [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+hook.Add( "CFC_Notifications_init", "settings_init", function()
+    CFCNotifications.clearSettingsCache()
+    -- displays a normal test notification that shows for 5 seconds
+    local notif = CFCNotifications.new( "test", "Text", true )
+    -- notif:SetPriority( 3 )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]: Remove or address?

File: lua/cfc_notifications/client/cl_settings.lua [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+            local t = setting.type
+            if type( def ) == "bool" then def = def and 1 or 0 end
+            CreateClientConVar( val, tostring( def ) )
+            cvars.AddChangeCallback( val, function( cvarName, old, new )
+                local cvar = GetConVar( cvarName )
+                if old == new then return end
+
+                -- Form:NumSlider auto changes values by removing trailing 0's, aka 0.60000 -> 0.60, this triggers a nocache
+                -- technically not a change, so lets ignore it
+                if t == "float" and tonumber( old ) == tonumber( new ) then return end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]: Very dense here - please liberate with additional whitespace

File: lua/cfc_notifications/client/cl_settings.lua [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+                c = panel:CheckBox( setting.displayName, val )
+            elseif setting.type == "action" then
+                c = panel:Button( setting.displayName, val )
+            elseif setting.type == "int" then
+                c = panel:NumSlider( setting.displayName, val, setting.min or 0, setting.max or 100, 0 )
+            elseif setting.type == "float" then
+                c = panel:NumSlider( setting.displayName, val, setting.min or 0, setting.max or 100, 2 )
+            elseif setting.type == "string" then
+                c = panel:TextEntry( setting.displayName, val )
+            end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/client/cl_settings.lua]: Similarly very dense, could be made nicer with some whitespace

File: lua/cfc_notifications/server/sv_net.lua [hunk 22]
```diff
@@ -0,0 +1,78 @@
+local function forceWriteTable( tab )
+    for k, v in pairs( tab ) do
+
+        local typeID = TypeID( v )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/server/sv_net.lua]:
```suggestion
    for k, v in pairs( tab ) do
        local typeID = TypeID( v )

```

File: lua/cfc_notifications/server/sv_net.lua [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        if typeID == TYPE_TABLE then
+            -- Ensure we recurse with this tablewrite, no the original
+            net.WriteType( k )
+            net.WriteUInt( typeID, 8 )
+            forceWriteTable( v )
+        elseif net.WriteVars[TypeID( v )] then
+            -- Just gonna skip over things we can't encode
+            net.WriteType( k )
+            net.WriteType( v )
+        end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/server/sv_net.lua]: Even with the comments, I'm having a hard time understanding what these conditions _mean_

File: lua/cfc_notifications/server/sv_net.lua [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+function CFCNotifications.Base:_callClient( ply, funcName, ... )
+    net.Start( "CFC_NotificationEvent" )
+    net.WriteString( self:GetID() )
+    net.WriteString( funcName )
+    net.WriteTable( { ... } )
+    net.Send( ply )
+end
+
+function CFCNotifications.Base:IsNotificationShowing( ply )
+    return #CFCNotifications.Base:GetPopupIDs( ply ) > 0
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/server/sv_net.lua:62]: You're sure `#` is okay here? Doesn't need `table.Count` ?

File: lua/cfc_notifications/server/sv_net.lua [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+end
+
+function CFCNotifications.Base:GetPopupIDs( ply )
+    if self._popupIDs and self._popupIDs[ply] then
+        return table.GetKeys( self._popupIDs[ply] )
+    end
+    return {}
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/server/sv_net.lua:70]:
```suggestion

    return {}
```

File: lua/cfc_notifications/shared/sh_base.lua [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+CFCNotifications.PRIORITY_MAX = 5
+
+-- naming convention:
+-- camelCase for methods on CFCNotifications
+-- UpperCamelCase for methods on a Notification object ( CONTEXT )
+-- prefix with _ for private
+
+-- includes + network strs
+include( "sh_context.lua" )
+include( "sh_helper.lua" )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua:18]:
```suggestion
include( "sh_helper.lua" )

```

File: lua/cfc_notifications/shared/sh_base.lua [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+
+    hook.Add( "InitPostEntity", "CFCNotifications_request_data", function()
+        -- Acting as a way of saying "I'm ready"
+        net.Start( "CFC_NotificationExists" )
+        net.SendToServer()
+    end )
+end
+
+function CFCNotifications.registerNotificationType( notificationType, callback, inherit )
+    local CONTEXT
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua:41]:
```suggestion
    local CONTEXT

```

File: lua/cfc_notifications/shared/sh_base.lua [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+    if inherit then
+        CONTEXT = CFCNotifications.Types[inherit]
+        if not CONTEXT then
+            error( "Unknown inherited notification type " .. inherit )
+        end
+    else
+        CONTEXT = CFCNotifications.Base
+    end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua:50]:
```suggestion
    end

```

File: lua/cfc_notifications/shared/sh_base.lua [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+        else
+            error( "Notification id " .. id .. " already in use. Pass true as third argument to force replace." )
+        end
+    end
+    local notif = {}
+    notif._id = id
+    notif._type = notificationType
+    local mt = {}
+    mt.__index = CONTEXT
+    setmetatable( notif, mt )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua]: Can we get more whitespace and better var names or more comments to clarify this section?

File: lua/cfc_notifications/shared/sh_base.lua [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+        filter = fWrap( filter )
+    end
+    filter = filter or player.GetAll
+    local players = filter()
+    if type( players ) == "Player" then
+        players = { players }
+    end
+    if not players then players = player.GetAll() end
+    return players
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua]:
```suggestion
function CFCNotifications._resolveFilter( filter )
    if type( filter ) == "Player" then
        filter = { filter }
    end
    
    if type( filter ) == "table" then
        filter = fWrap( filter )
    end
    
    filter = filter or player.GetAll
    local players = filter()
    
    if type( players ) == "Player" then
        players = { players }
    end
    
    if not players then players = player.GetAll() end
    
    return players
end
```

File: lua/cfc_notifications/shared/sh_base.lua [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+hook.Add( "CFC_Notifications_init", "init_message", function()
+    print( "CFCNotifications loaded." )
+end )
+
+function CFCNotifications.reload()
+    -- Escape this context
+    timer.Simple( 0, function()
+        hook.Run( "CFC_Notifications_stop" )
+        include( "cfc_notifications/shared/sh_base.lua" )
+        -- Wait for new notifs to load
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_base.lua:161]:
```suggestion

        -- Wait for new notifs to load
```

File: lua/cfc_notifications/shared/sh_context.lua [hunk 32]
```diff
@@ -0,0 +1,193 @@
+CFCNotifications._timerNameCount = 1
+CFCNotifications.contextHelpers = {}
+
+function CFCNotifications._getTimerName()
+    local newName = "notification-timer-" .. tostring( CFCNotifications._timerNameCount )
+
+    CFCNotifications._timerNameCount = CFCNotifications._timerNameCount + 1 -- How I yearn for the ++ operator
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua]:
```suggestion
    CFCNotifications._timerNameCount = CFCNotifications._timerNameCount + 1
```

😭 😭 👶

File: lua/cfc_notifications/shared/sh_context.lua [hunk 33]
```diff
@@ -0,0 +1,193 @@
...
+
+    return newName
+end
+
+local function checkTypes( tab, t )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua]: `isOfType` ?

File: lua/cfc_notifications/shared/sh_context.lua [hunk 34]
```diff
@@ -0,0 +1,193 @@
...
+    for k, v in pairs( tab ) do
+        if type( v ) ~= t then
+            return false
+        end
+    end
+    return true
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua:19]:
```suggestion

    return true
```

File: lua/cfc_notifications/shared/sh_context.lua [hunk 35]
```diff
@@ -0,0 +1,193 @@
...
+        end
+        net.Start( "CFC_NotificationExists" )
+        net.WriteString( self:GetID() )
+        net.WriteBool( false )
+        net.Broadcast()
+    else
+        self:RemovePopups()
+        CFCNotifications._reloadIgnoredPanels()
+    end
+    CFCNotifications.Notifications[self:GetID()] = nil
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua:39]:
```suggestion

    CFCNotifications.Notifications[self:GetID()] = nil
```

File: lua/cfc_notifications/shared/sh_context.lua [hunk 36]
```diff
@@ -0,0 +1,193 @@
...
+        -- Lets not pop up the same notification more than once
+        if not self:GetAllowMultiple() and self:IsNotificationShowing( self ) then
+            return
+        end
+
+        -- Check it's allowed to show
+        if self:GetCloseable() and not self:ShouldShowNotification() then
+            return
+        end
+        -- Call something in render.lua to make a panel for this notif
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua]:
```suggestion

        -- TODO: Call something in render.lua to make a panel for this notif
```

is this comment still relevant?

File: lua/cfc_notifications/shared/sh_context.lua [hunk 37]
```diff
@@ -0,0 +1,193 @@
...
+end
+
+function CONTEXT:SendDelayed( delay, filter )
+    self:SendRepeated( delay, 1, filter )
+end
+
+function CONTEXT:SendRepeated( delay, reps, filter )
+    if self._timerName then
+        error( "Timer already running" )
+    end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua:88]:
```suggestion
    end

```

File: lua/cfc_notifications/shared/sh_context.lua [hunk 38]
```diff
@@ -0,0 +1,193 @@
...
+    self._timerName = CFCNotifications._getTimerName()
+    timer.Create( self._timerName, delay, reps, function()
+        self:Send( filter )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua:92]:
```suggestion
        self:Send( filter )

```

File: lua/cfc_notifications/shared/sh_context.lua [hunk 39]
```diff
@@ -0,0 +1,193 @@
...
+        elseif type( v ) ~= argType then
+            error( "Unexpected type in Set" .. externalName .. ", expected " .. argType .. ", got " .. type( v ) )
+        end
+        self[internalName] = v
+        if onChange then onChange( self, v ) end
+    end
+    context["Get" .. externalName] = function( self )
+        return self[internalName]
+    end
+end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_context.lua]: This part is fairly dense, too - can you think of any ways we could slim it down? (also whitespace? 😄 )

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 40]
```diff
@@ -0,0 +1,117 @@
...
+    notif:SetCloseable( false )
+    notif:SetPriority( CFCNotifications.PRIORITY_MAX )
+    notif:Send( filter )
+end
+
+if SERVER then
+    local function plural( n )
+        if n == 1 then return "" end
+        return "s"
+    end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua:32]:
```suggestion

        return "s"
    end

```

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 41]
```diff
@@ -0,0 +1,117 @@
...
+    function CFCNotifications.startVote( id, question, time, options, cb )
+        local votes = {}
+        local ended = false
+        local notif = CFCNotifications.new( id, "Buttons", true )
+        notif:SetText( question )
+        notif:SetAlwaysTiming( true )
+        notif:SetDisplayTime( time or 30 )
+        notif:SetTimed( true )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua]:
```suggestion
        local votes = {}
        local ended = false
        local notif = CFCNotifications.new( id, "Buttons", true )
        
        notif:SetText( question )
        notif:SetAlwaysTiming( true )
        notif:SetDisplayTime( time or 30 )
        notif:SetTimed( true )
```

or

```suggestion
        local votes = {}
        local ended = false
        
        local notif = CFCNotifications.new( id, "Buttons", true )
        notif:SetText( question )
        notif:SetAlwaysTiming( true )
        notif:SetDisplayTime( time or 30 )
        notif:SetTimed( true )
```

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 42]
```diff
@@ -0,0 +1,117 @@
...
+        local plyTotal = 0
+        local plyCount = 0
+        local voteCount = 0
+
+        for k, v in pairs( options or {} ) do
+            notif:AddButton( v, Color( 255, 255, 255 ), v )
+        end
+
+        local function endVote()
+            ended = true
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua:52]: Global `ended` ?

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 43]
```diff
@@ -0,0 +1,117 @@
...
+            local winners = {}
+            local maxVote = 0
+            for k, v in pairs( votes ) do
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua]:
```suggestion
            local maxVote = 0

            for k, v in pairs( votes ) do
```

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 44]
```diff
@@ -0,0 +1,117 @@
...
+                    "Option" .. plural( #winners ) .. " \"" .. table.concat( winners, "\", \"" ) ..
+                    "\" won with " .. maxVote .. " vote" .. plural( maxVote ) .. "." )
+            end
+
+            if cb then
+                if #winners == 1 then
+                    if not options then
+                        winners[1] = winners[1] == "Yes"
+                    end
+                    cb( winners[1] )
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua]:
```suggestion
	                    end

                    cb( winners[1] )
```

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 45]
```diff
@@ -0,0 +1,117 @@
...
+            if ended then return end
+            if type( option ) == "boolean" then option = option and "Yes" or "No" end
+            votes[option] = votes[option] or 0
+            votes[option] = votes[option] + 1
+            plyCount = plyCount + 1
+            voteCount = voteCount + 1
+            if plyCount == plyTotal then
+                notif:Remove()
+                timer.Simple( 1, endVote )
+            end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua]:
```suggestion
            if ended then return end
            if type( option ) == "boolean" then option = option and "Yes" or "No" end

            votes[option] = votes[option] or 0
            votes[option] = votes[option] + 1

            plyCount = plyCount + 1
            voteCount = voteCount + 1

            if plyCount == plyTotal then
                notif:Remove()
                timer.Simple( 1, endVote )
            end
```

File: lua/cfc_notifications/shared/sh_helper.lua [hunk 46]
```diff
@@ -0,0 +1,117 @@
...
+        end
+
+        function notif:OnClose( ply )
+            if ended then return end
+            plyCount = plyCount + 1
+            if plyCount == plyTotal then
+                notif:Remove()
+                timer.Simple( 1, endVote )
+            end
+        end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_helper.lua]:
```suggestion
        function notif:OnClose( ply )
            if ended then return end

            plyCount = plyCount + 1

            if plyCount == plyTotal then
                notif:Remove()
                timer.Simple( 1, endVote )
            end
        end
```

File: lua/cfc_notifications/shared/sh_presets.lua [hunk 47]
```diff
@@ -0,0 +1,139 @@
...
+
+    function CONTEXT:AddButton( text, col, ... )
+        col = col or Color( 255, 255, 255 )
+        self._buttons = self._buttons or {}
+        table.insert( self._buttons, {
+            text = text,
+            color = col,
+            data = { ... }
+        } )
+    end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_presets.lua]:
```suggestion
    function CONTEXT:AddButton( text, col, ... )
        col = col or Color( 255, 255, 255 )
        self._buttons = self._buttons or {}

        table.insert( self._buttons, {
            text = text,
            color = col,
            data = { ... }
        } )
    end
```

File: lua/cfc_notifications/shared/sh_presets.lua [hunk 48]
```diff
@@ -0,0 +1,139 @@
...
+                -- Delay the close to see the button animation, make it clear that the button is what was selected
+                timer.Simple( 0.5, function()
+                    CFCNotifications._removePopup( panel )
+                    this:_callHook( popupID, "OnButtonPressed", unpack( btnData.data or { btnData.text } ) )
+                end )
+            end
+            btn:SetSize( btnW - 20, 30 )
+            btn:SetPos( 10 + ( k - 1 ) * btnW, h - 40 )
+            table.insert( btns, btn )
+        end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_presets.lua]: This section could also use a little stylistic love. Also a good opportunity to extract some magic numbers and calculations

File: lua/cfc_notifications/client/cl_customalphato.lua [hunk 49]
```diff
@@ -0,0 +1,44 @@
+local fadePanels = {}
+
+hook.Add( "Think", "ca_think", function()
```
Comment (plally) [inline lua/cfc_notifications/client/cl_customalphato.lua]: perhaps a better hook name could be used here

File: lua/cfc_notifications/shared/sh_presets.lua [hunk 50]
```diff
@@ -0,0 +1,139 @@
...
+                self._unhoverTime = nil
+                self._targetAlpha = 100
+                if not self._hidden then
+                    self:CustomAlphaTo( self._targetAlpha, 0.5 )
+                end
+            end
+            panel._lastHovered = panel._hovered
+
+            oldThink( self )
+        end
```
Comment (sarahsturgeon) [inline lua/cfc_notifications/shared/sh_presets.lua]: Does this spacing look any better to you? I like to have newlines between defining setup variables and the conditions/actions

```suggestion
    function CONTEXT:PopulatePanel( canvas, popupID, panel )
        panel._unhoverTime = SysTime() + 1
        local oldThink = panel.Think

        function panel:Think()
            local x, y = self:LocalCursorPos()
            local w, h = self:GetSize()
            self._hovered = not ( x < 0 or x > w or y < 0 or y > h )

            if self._hovered ~= self._lastHovered then
                if self._hovered then
                    self._targetAlpha = 255

                    if not self._hidden then
                        self:CustomAlphaTo( self._targetAlpha, 0.1 )
                    end
                else
                    self._unhoverTime = SysTime()
                end
            end

            if not self._hovered and self._unhoverTime and SysTime() - self._unhoverTime > 1 then
                self._unhoverTime = nil
                self._targetAlpha = 100

                if not self._hidden then
                    self:CustomAlphaTo( self._targetAlpha, 0.5 )
                end
            end

            panel._lastHovered = panel._hovered

            oldThink( self )
        end
```

**Review by sarahsturgeon — COMMENTED** [review]
Here's a 100% stylistic review, focusing mainly on opportunities to improve spacing.

We already went through the structure on a video call and I thought it looked good.

File: README.md (hunk 1 above)
Comment (samuelWilliams99) [reply README.md]: They can do both, but I found the context menu puts a gui in front of notifications, annoyingly

File: lua/cfc_notifications/client/cl_dnotification.lua (hunk 3 above)
Comment (samuelWilliams99) [reply lua/cfc_notifications/client/cl_dnotification.lua]: Well `w` and `h` are dependent on Paint. I could perhaps pull `frac` out into `Think`?

File: lua/cfc_notifications/client/cl_dnotification.lua (hunk 4 above)
Comment (samuelWilliams99) [reply lua/cfc_notifications/client/cl_dnotification.lua]: Are we actually using CFC color lib? What's the process of that, do I just include the whole thing and add my color to it?

File: lua/cfc_notifications/server/sv_net.lua (hunk 24 above)
Comment (samuelWilliams99) [reply lua/cfc_notifications/server/sv_net.lua:62]: Below in `GetPopupIDs`, the return is just `table.GetKeys` or `{}`, both of which are sequential, and so `#` is fine.

File: lua/cfc_notifications/shared/sh_helper.lua (hunk 42 above)
Comment (samuelWilliams99) [reply lua/cfc_notifications/shared/sh_helper.lua:52]: It's local in the containing function

**Comment by samuelWilliams99** [comment]
The omega review has been defeated. Rejoice!

**Review by sarahsturgeon — APPROVED** [review]
LGTM
